### PR TITLE
fix: MultiSteps with mixed precision params and updates.

### PR DESCRIPTION
**Pull request by hlzl**
Aims at solving #377 and #1038 

## Problem
`MultiSteps` throws the following error when using different precision for parameters and the corresponding gradient values (e.g. `bfloat16` parameters but casting the parameters' gradient to `float32` for scaling such as in `scale_by_factored_rms` and `flax.training.dynamic_scale`):
```bash
File "/home/user/miniforge3/envs/name/lib/python3.12/site-packages/optax/transforms/_accumulation.py", line 380, in update
    new_updates, new_state = lax.cond(
                             ^^^^^^^^^
TypeError: true_fun and false_fun output must have identical types, got
({'decoder_image_output': {'bias': 'DIFFERENT ShapedArray(bfloat16[768]) vs. ShapedArray(float32[768])', 'kernel': ...
```
This is due to `jax.lax.cond` expecting the outputs of both conditions to have the same `dtype`.

## Code to Reproduce
Taken from #377 :
```python
import os; os.environ['JAX_PLATFORMS'] = 'cpu'
import jax
import jax.numpy as jnp
import optax

@jax.jit
@jax.value_and_grad
def f(params, x, labels):
    logits = params @ x
    loss = optax.softmax_cross_entropy_with_integer_labels(logits, labels)
    return loss.mean()

params = jnp.zeros((5, 18), dtype=jnp.bfloat16)
x = jnp.zeros((18, 4), dtype=jnp.bfloat16)
labels = jnp.zeros((5,), dtype=jnp.uint16)
value, grad = f(params, x, labels)

lr = 0.00005
n_accumulation_steps = 4

optimizer = optax.adafactor(learning_rate=lr)
optimizer = optax.MultiSteps(optimizer, n_accumulation_steps)
opt_state = optimizer.init(params)
updates, opt_state = optimizer.update(grad, opt_state, params)
print(updates)
```

## Proposed Fix
Casting new updates to the `dtype` of the original optimizer state. Allows to solve the problem independent of the update `dtype`.

### CODE REVIEW

File: optax/transforms/_accumulation.py [hunk 1]
```diff
@@ -334,7 +334,9 @@ def update(self,
     # Note: we do not enclose variables to allow JAX to re-use memory buffers.
     def _do_update(updates, state, params):
       acc_grads = jtu.tree_map(
-          lambda upd, acc: self._acc_update(upd, acc, n_acc=state.mini_step),
+          lambda upd, acc: self._acc_update(
```
Comment (hlzl) [inline optax/transforms/_accumulation.py:337]: Now that I look at it again, a slightly better version would be the following as we can expect the leaves to be `jnp.array`s:
```python
acc_grads = jtu.tree_map(
          lambda upd, acc: self._acc_update(upd.astype(acc.dtype), acc, n_acc=state.mini_step)
          updates,
          state.acc_grads,
      )
```
and the following for line 365 ff.:
```python
final_updates = jtu.tree_map(
           lambda ga, st: emit * ga.astype(st.dtype),
           final_updates,
           state.acc_grads
       )
```
Let me know what you think of the proposed fix in general and I'll updates these.

**Comment by vroulet** [comment]
Thank you very much for tacking a look at this!
1. Could you add a test?
2. Since the goal is to ensure the same types whether the update is actually used or not, one may use `jax.eval_shape`.

Happy to take a closer look once you have a test ready to fiddle with different solutions.

**Comment by vroulet** [comment]
Hello @hlzl,

Thank you again for looking into this! Looking back at your example the only culprit is in the definition of the `zero_updates` in the `_skip_update` function.
Thank you again for the MWE, below is a proposed fix.
Would you be willing to do this and add a test?

Thank you again!
```
import os; os.environ['JAX_PLATFORMS'] = 'cpu'
import jax
import jax.numpy as jnp
import jax.tree_util as jtu
import optax
import optax.tree_utils as otu

@jax.jit
@jax.value_and_grad
def f(params, x, labels):
    logits = params @ x
    loss = optax.softmax_cross_entropy_with_integer_labels(logits, labels)
    return loss.mean()

params = jnp.zeros((5, 18), dtype=jnp.bfloat16)
x = jnp.zeros((18, 4), dtype=jnp.bfloat16)
labels = jnp.zeros((5,), dtype=jnp.uint16)
value, grad = f(params, x, labels)

lr = 0.00005
n_accumulation_steps = 4

optimizer = optax.adafactor(learning_rate=lr)
optimizer = optax.MultiSteps(optimizer, n_accumulation_steps)
state = optimizer.init(params)
# updates, opt_state = optimizer.update(grad, opt_state, params)

extra_args = {}
zero_updates = otu.tree_zeros_like(state.acc_grads)
print(f'zero updates dtype before fix: {jtu.tree_map(jnp.dtype, zero_updates)}' )
update_shape, _ = jax.eval_shape(
    optimizer.inner_opt.update,
    updates=otu.tree_zeros_like(params),
    state=state.inner_opt_state,
    params=otu.tree_zeros_like(params),
    **extra_args
)
zero_updates = jtu.tree_map(lambda x, s: x.astype(s.dtype), zero_updates, update_shape)
print(f'zero updates dtype after fix: {jtu.tree_map(jnp.dtype, zero_updates)}' )
```

**Comment by hlzl** [comment]
Hi @vroulet,
Thank you for looking into the fix and sorry for my late responds!
I'm happy to have a look and add the test. Unfortunately, I'm a bit busy the next 2-3 weeks. I'll try to find some time for this in-between but will definitely have a look after my deadline. Hope this works for you!

**Comment by vroulet** [comment]
No worries, it can wait, thank you again for looking into that!

**Comment by mar-muel** [comment]
Hello - I'm also being blocked by this - Great to see a fix is in the workings 🙏 

In the meantime, is there a workaround or is `optax.MultiSteps` just generally incompatible with non-fp32 params for now?

**Comment by vroulet** [comment]
@mar-muel, as a quick patch you may try adding just before calling the update function.
```
update_dtype = ... # add the expected dtype of the optimizer you are working with
state = optax.tree_utils.tree_set('acc_grads', optax.tree_utils.tree_cast(state.acc_grads, update_dtype)
```
PS: I don't think that optax.MultiSteps is generally incompatible with non-fp32 params. @hlzl singled out well the issue I think (mismatch between type of returned updates of inner optimizer and type of generic gradients (which is normally the same type as the parameters). @mar-muel are you referring to another issue?

**Comment by mar-muel** [comment]
Unfortunately, there seems to be another upcasting happening internally in optax. So even if one manually casts the `inner_opt_state` to bfloat16 before applying the updates the same error will be raised (`DIFFERENT ShapedArray(bfloat16[3072]) vs. ShapedArray(float32[3072])`).

For reference here is what I've tried to do:
```python
inner_opt_state = jax.tree_util.tree_map(lambda s: s.astype(jnp.bfloat16) if s.dtype == jnp.float32 else s, opt_state.inner_opt_state)
opt_state = opt_state._replace(inner_opt_state=inner_opt_state)       

# optimizer.update(...)                                                                                                                               
```

**Comment by hlzl** [comment]
@mar-muel I believe casting the update to be of the opt_state type is more robust. You could try copying my proposed changes from the PR and update your local optax version (it's only a few lines of code). A bit of a hack for right now but would also allow to test my changes ;)

**Comment by mar-muel** [comment]
Hey @hlzl - Yes, I can confirm that the fix works (I've applied your changes to my current `optax==0.2.2`). However, I run into OOM sadly. Is the opt state still allocated in fp32 at one point, yielding a memory spike? It also seems like optax.MultiSteps is not very memory efficient in general, given that it keeps a full duplicate of the opt state :/

**Comment by hlzl** [comment]
@mar-muel Good to know, thanks! 
As far as I can tell, all states within the optimizer (which is `inner_opt_state and `acc_grads`) are of the same types as the params you pass to the optimizer when calling `init`, no other upcasting.

**Comment by vroulet** [comment]
The culprit is not MultiSteps. The culprits are all optimizers that do not preserve the dtype of the parameters. I'm working on a PR to correct that across the code base. The PR should be ready today and pushed next week. I'll ping you then to know if you still encounter issues after the pass.

**Comment by vroulet** [comment]
#1060 solves #1038 and #337, see tests, so I think this PR may be closed once #1060 is merged.
Thank you again @hlzl for investigating this!
The optimizers that posed problems were adafactor, sm3 and novograd in the alias file, and most of the optimizers in the contrib file.
@mar-muel, what optimizer were you considering?
About
> It also seems like optax.MultiSteps is not very memory efficient in general, given that it keeps a full duplicate of the opt state :/
if you have some probing tools to assert that, it would be great. HLO is usually doing a very good job at reusing buffers when possible, but it indeed seems that ` state.inner_opt_state` and  `new_inner_state` may at one point occupy two different buffers in this code. If you don't have such probing tools, posting an issue would be great. Thanks for all!

**Comment by mar-muel** [comment]
@vroulet Indeed the problem seems to only occur with adafactor, not with adam (using optax==0.2.2, without any patches)

**Comment by vroulet** [comment]
@hlzl, @mar-muel, I'm closing this PR since #1060 has been merged and solved aforementioned issues.
Feel free to reopen it if I'm mistaken.

**Comment by hlzl** [comment]
Hi @vroulet , great to see the updates on the optimizer-side in optax! 

I would still propose to additionally introduce the checks and conversions from this PR in `MultiSteps` as there might be differences in the dtypes due to external libraries. E.g. I had this problem originally with the dynamic scaler of flax.
You can adapt the original example to reproduce the problem when using flax' `dynamic_scale` as follows:
```python
import os

os.environ["JAX_PLATFORMS"] = "cpu"
import jax
import jax.numpy as jnp
import optax
from flax.training.dynamic_scale import DynamicScale


@jax.jit
def f(params, x, labels):
    logits = params @ x
    loss = optax.softmax_cross_entropy_with_integer_labels(logits, labels)
    return loss.mean()


dyn_scale = DynamicScale(). # changed
params = jnp.zeros((5, 18), dtype=jnp.bfloat16)
x = jnp.zeros((18, 4), dtype=jnp.bfloat16)
labels = jnp.zeros((5,), dtype=jnp.uint16)
dynamic_scale, is_fin, loss, grad = dyn_scale.value_and_grad(f)(params, x, labels). # changed

lr = 0.01
n_accumulation_steps = 4

optimizer = optax.sgd(learning_rate=lr)
optimizer = optax.MultiSteps(optimizer, n_accumulation_steps)
opt_state = optimizer.init(params)
updates, opt_state = optimizer.update(grad, opt_state, params)
print(updates)
```

Let me know if we should thus reopen this PR or keep as is.

**Comment by vroulet** [comment]
Hello @hlzl !
Beyond the bug in the MultiSteps function, what we want to avoid is having dtype promotions when applying the optimizer. In other words, a code like
```
grads = jax.grad(loss)(params)
updates, _ = opt.update(grads, state, params)
params = optax.apply_update(params, updates)
```
shall not induce in its last line a promotion of params.dtype. To me, any such piece of code that induces a promotion is a bug, and actually MultiSteps is catching somehow this bug.

However, optax should be able to handle mixed precision training. So in the code above, `optax.apply_update` can be modified while keeping dtypes of updates and params different. And all optimizers could also allow for dtypes that could be different of the dtypes of the params but we need to be careful to document such an approach (so that the user knows exactly that there may be a dtype promotion) with an adequate flag.

To conclude, I see what you mean, and we may incorporate such a change. But there is a more important underlying issue that needs to be documented (about mixed precision training). So if you could document clearly such an issue after reviewing some solutions like the ones discussed in #336, it would be great.

**Comment by hlzl** [comment]
@vroulet I'm not sure I understand what you mean. 

Currently, in the case with the dynamic scaler it is not possible to run `optimizer.update(grad, opt_state, params)` even before applying any updates because `MultiSteps` is causing the same error as from #377 and #1038 .

The fix doesn't change the dtype of `params`, instead it adapts `grad` to be of the same dtype as `params` (which is necessary to run  `optimizer.update(...)` when using `MultiSteps`).
`updates` is then already of the same dtype as `params`.

E.g. adding the following loop to end of the previous example would fail in the first line without the fix but consistently returns `bfloat16` when using the proposed fix:
```python
for e in range(n_accumulation_steps * 2):
    updates, opt_state = optimizer.update(grad, opt_state, params)
    params = optax.apply_updates(params, updates)
    print(f"Step {e}: grad {grad.dtype} updates {updates.dtype} - params {params.dtype}") 
```

Note that this is a `MultiSteps` specific behavior. Commenting `MultiSteps` out, the script runs without an issue and returns `grad` and `updates` as `float32` while `params` is returned as `bfloat16`.

**Comment by hlzl** [comment]
Ah, I see now. Is it possible that you mixed up the order of `updates` and `params` in `optax.apply_updates`? I think it should be `optax.apply_updates(params, updates)`, which also results in no promotion.

Other than that, I still think `updates` should be of the same dtype independent of `MultiSteps` being used or not. So my current proposed fix is not clean. I think a better alternative is to change how we initialize `updates` within `MultiSteps`. 

Let me know if changing the order of the arguments inside `apply_updates` solved your issue and I'll look into this.

**Comment by vroulet** [comment]
Ah sorry indeed by setting `optax.apply_updates(params, updates)` it works. 
So yes, MultiSteps still has an issue I agree. It should mimic the behavior of apply_updates.

Sorry for the long discussion but this helps too! I think #1060 should probably be patched in light of that.
Namely, we do not necessarily want the updates of the type of params but we probably want them of the same dtype as the gradients. I'll open an issue to make sure #1060 aligns with this.

Thank you again! So you can reopen this PR or create a new one (I let you choose) with your fix and add corresponding tests.

**Comment by vroulet** [comment]
Thanks @hlzl  for following up on this! First, we could schedule a meeting to chat about this directly, it could speed things up. Send me an email and let's schedule it.

> the update and more importantly the states (e.g. mu and nu) are always promoted to float32.

What are (i) the dtype of params, (ii) the dtype of gradients? Currently, the attributes of the state follow the dtype of the params. If `params_dtype` < `grads_dtype` (with < being defined by dtype promotion see e.g. the `tree_dtype` function for more explanation if it's not clear), the attributes of the state get promoted to the type of the grads after the first update (not ideal because it induces a recompilation). If `grads_dtype` < `params_dtype`, the dtype of the attributes may never match the dtype of the gradients (and the dtype of gradients will always be promoted to the dtype of params). I could make a whole pass on the codebase to prevent that. But so I'd like to know your use case.

Generally, this is a design issue of the init function of the gradient transformations (we should a priori give "fake" gradients rather than params to the init function). Mostly, it may be a simple renaming of the inputs in the init function for the user to understand that the init function should rather be fed gradients-like variables rather than params-like parameters. There may be some cases where parameters should still be fed and in that case I think one may consider modifying the init function to accept some additional arguments (just as GradientTransformationExtraArgs). This would then be a change in API that requires careful handling. But so again chatting together about your use case would be helpful (I have not considered mixed precision training in my work so some inputs from you will be great).

> Do you have an idea how I can change the dtype of states within e.g. ScaleByAdamState without doing a full update and increasing the count @vroulet ?

One could simply use a tree_map to convert dtypes as long as the leaves match the structure of the params. Where would you need that exactly?

> Maybe we can discuss here and I'll open a new one once it works as planned.

Sounds good, again could be good to discuss that "de vive voix"

**Comment by hlzl** [comment]
@vroulet A meeting sounds good, I'll send you an email!

> we should a priori give "fake" gradients rather than params to the init function

This basically summarizes where the problems come from. Dealing with this is quite straight forward for most cases but a bit tricky when the dtypes change in the `inner_opt_state` in one of the two conditions (see promotion in https://github.com/hlzl/optax/blob/73a19a266f6d1aa6a61ff480bd30a2367a39c79a/optax/transforms/_accumulation.py#L344 vs. no change in https://github.com/hlzl/optax/blob/73a19a266f6d1aa6a61ff480bd30a2367a39c79a/optax/transforms/_accumulation.py#L377).

**Pull request closed without merging**